scvi/external/scar/_model.py

```diff
@@ -0,0 +1,334 @@
+        scale_activation: Literal["softmax", "softplus", "softplus_sp"] = "softplus_sp",
+        sparsity: float = 0.9,
+        **model_kwargs,
+    ):
+        super(SCAR, self).__init__(adata)
+
+        n_batch = self.summary_stats.n_batch
+        use_size_factor_key = (
+            REGISTRY_KEYS.SIZE_FACTOR_KEY in self.adata_manager.data_registry
+        )
+        library_log_means, library_log_vars = None, None
+        if not use_size_factor_key:
+            library_log_means, library_log_vars = _init_library_size(
+                self.adata_manager, n_batch
+            )
+
+        # self.summary_stats provides information about anndata dimensions and other tensor info
+        if not torch.is_tensor(ambient_profile):
+            if isinstance(ambient_profile, str):
+                ambient_profile = np.nan_to_num(adata.varm[ambient_profile])
+            elif isinstance(ambient_profile, pd.DataFrame):
+                ambient_profile = ambient_profile.fillna(0).values
+            elif isinstance(ambient_profile, np.ndarray):
+                ambient_profile = np.nan_to_num(ambient_profile)
+            elif not ambient_profile:
+                ambient_profile = adata.X.sum(axis=0) / adata.X.sum(axis=0).sum()
+                ambient_profile = np.nan_to_num(ambient_profile)
+            else:
+                raise TypeError(
+                    f"Expecting str / np.array / None / pd.DataFrame, but get a {type(ambient_profile)}"
+                )
+            ambient_profile = (
+                torch.from_numpy(np.asarray(ambient_profile)).float().reshape(1, -1)
+            )
+
+        self.module = SCAR_VAE(
+            ambient_profile=ambient_profile,
+            n_input=self.summary_stats.n_vars,
+            n_hidden=n_hidden,
+            n_latent=n_latent,
+            n_layers=n_layers,
+            dropout_rate=dropout_rate,
+            gene_likelihood=gene_likelihood,
+            latent_distribution=latent_distribution,
+            use_size_factor_key=use_size_factor_key,
+            library_log_means=library_log_means,
+            library_log_vars=library_log_vars,
+            scale_activation=scale_activation,
+            sparsity=sparsity,
+            **model_kwargs,
+        )
+        self._model_summary_string = (
+            "SCVI-AR Model with the following params: \nn_hidden: {}, n_latent: {}, n_layers: {}, dropout_rate: "
+            "{}, gene_likelihood: {}, latent_distribution: {}"
+        ).format(
+            n_hidden,
+            n_latent,
+            n_layers,
+            dropout_rate,
+            gene_likelihood,
+            latent_distribution,
+        )
+        self.init_params_ = self._get_init_params(locals())
+
+    @classmethod
+    @setup_anndata_dsp.dedent
+    def setup_anndata(
+        cls,
+        adata: AnnData,
+        layer: Optional[str] = None,
+        batch_key: Optional[str] = None,
+        labels_key: Optional[str] = None,
```

Comment:
can't these be removed from here but kept internally in the function with None arguments?

Reply:
done in latest commit